File: bridge/python/src/destack/protocol/connection/embedded.py
from __future__ import annotations

from collections import deque
from collections.abc import Sequence
from typing import Protocol

from destack.protocol.codec import decode_payload_frame, encode_payload_frame

from .transport import TransportError


class EmbeddedServer(Protocol):
    """Server that can dispatch one workspace protocol payload."""

    def dispatch(self, data: bytes) -> Sequence[bytes]:
        """Dispatch one encoded protocol message payload."""


class EmbeddedTransport:
    """Embedded transport backed by an in-process workspace server."""

    def __init__(self, server: EmbeddedServer) -> None:
        self._server = server
        self._queue: deque[bytes] = deque()
        self._closed = False

    def send(self, data: bytes) -> None:
        """Send one binary frame."""

        if self._closed:
            raise TransportError("embedded transport is closed")

        payload = decode_payload_frame(data)
        responses = self._server.dispatch(payload)
        self._queue.extend(encode_payload_frame(response) for response in responses)

    def receive(self) -> bytes:
        """Receive one binary frame."""

        try:
            return self._queue.popleft()
        except IndexError as error:
            raise TransportError("embedded transport has no pending frame") from error

    def close(self) -> None:
        """Close the transport."""

        self._closed = True
        self._queue.clear()


__all__ = [
    "EmbeddedServer",
    "EmbeddedTransport",
]

File: client/typescript/src/protocol/connection/embedded.ts
import { decodePayloadFrame, encodePayloadFrame } from "../codec.js";
import { TransportError, type Transport } from "./transport.js";

/** Server that can dispatch one workspace protocol payload. */
export interface EmbeddedServer {
  /** Dispatch one encoded protocol message payload. */
  dispatch(bytes: Uint8Array): readonly unknown[];
}

/** Embedded transport backed by an in-process workspace server. */
export class EmbeddedTransport implements Transport {
  readonly #server: EmbeddedServer;
  readonly #queue: Uint8Array[] = [];
  readonly #waiters: {
    readonly resolve: (bytes: Uint8Array) => void;
    readonly reject: (error: TransportError) => void;
  }[] = [];
  #closed: TransportError | undefined;

  /** Create an embedded transport over one in-process server. */
  constructor(server: EmbeddedServer) {
    this.#server = server;
  }

  /** Send one binary frame. */
  async send(bytes: Uint8Array): Promise<void> {
    if (this.#closed !== undefined) {
      throw this.#closed;
    }

    const payload = decodePayloadFrame(bytes);
    const responses = this.#server.dispatch(payload);
    for (const response of responses) {
      this.#enqueue(encodePayloadFrame(this.#frameBytes(response)));
    }
  }

  /** Receive one binary frame. */
  receive(): Promise<Uint8Array> {
    if (this.#queue.length > 0) {
      return Promise.resolve(this.#queue.shift()!);
    }

    if (this.#closed !== undefined) {
      return Promise.reject(this.#closed);
    }

    return new Promise((resolve, reject) => {
      this.#waiters.push({ resolve, reject });
    });
  }

  /** Close the transport. */
  close(): void {
    if (this.#closed !== undefined) {
      return;
    }

    this.#closed = new TransportError("embedded transport closed");
    this.#queue.length = 0;
    while (this.#waiters.length > 0) {
      const waiter = this.#waiters.shift()!;
      waiter.reject(this.#closed);
    }
  }

  #enqueue(bytes: Uint8Array): void {
    const waiter = this.#waiters.shift();

    if (waiter !== undefined) {
      waiter.resolve(bytes);
    } else {
      this.#queue.push(bytes);
    }
  }

  #frameBytes(frame: unknown): Uint8Array {
    if (frame instanceof Uint8Array) {
      return frame;
    }

    if (frame instanceof ArrayBuffer) {
      return new Uint8Array(frame);
    }

    if (ArrayBuffer.isView(frame)) {
      return new Uint8Array(frame.buffer, frame.byteOffset, frame.byteLength);
    }

    if (Array.isArray(frame)) {
      return new Uint8Array(frame);
    }

    throw new TransportError("embedded transport frame was not binary");
  }
}
